include/revng/Model/Binary.h

```diff
@@ -6,601 +6,67 @@
 
 #include "llvm/ADT/SmallString.h"
 
-#include "revng/ADT/KeyedObjectContainer.h"
-#include "revng/ADT/MutableSet.h"
-#include "revng/ADT/SortedVector.h"
-#include "revng/ADT/UpcastablePointer.h"
-#include "revng/ADT/UpcastablePointer/YAMLTraits.h"
-#include "revng/Model/Register.h"
-#include "revng/Model/TupleTree.h"
-#include "revng/Model/Type.h"
+#include "revng/Model/DynamicFunction.h"
+#include "revng/Model/Function.h"
+#include "revng/Model/Segment.h"
+#include "revng/Model/VerifyHelper.h"
 #include "revng/Support/MetaAddress.h"
-#include "revng/Support/MetaAddress/YAMLTraits.h"
-#include "revng/Support/YAMLTraits.h"
 
-// Forward declarations
-namespace model {
-class VerifyHelper;
-class Function;
-class DynamicFunction;
-class Binary;
-class FunctionEdge;
-class CallEdge;
-class BasicBlock;
-class Segment;
-} // namespace model
+/* MODEL-YAML
+name: Binary
+doc: Data structure representing the whole binary
+type: struct
+fields:
+  - name: Functions
+    doc: List of the functions within the binary
+    array:
+      type: SortedVector
+      elementType: model::Function
+  - name: ImportedDynamicFunctions
+    doc: List of the functions within the binary
+    array:
+      type: SortedVector
+      elementType: model::DynamicFunction
+  - name: Architecture
+    doc: Binary architecture
+    type: model::Architecture::Values
+  - name: Segments
+    doc: List of segments in the original binary
+    array:
+      type: SortedVector
+      elementType: model::Segment
+  - name: EntryPoint
+    doc: Program entry point
+    type: MetaAddress
+  - name: Types
+    doc: The type system
+    # TODO: recognize this is upcastable
```

Comment:
I meant that it is possible to parse `model::Type` and drop the need to specify `upcastable: true`. I'm not sure we want to do it, because maybe you do want a list of `model::Type` and not a list of `UpcastablePointer<model::Type>`.

Reply:
Mmmh, having `SortedVector<UpcastablePointer<model::Type>>` vs `SortedVector<model::Type>` is an orthogonal problem.

I think yo use `upcastable: true` to emit some code to make this usable into an `UpcastablePointer`, isn't it?
If so, we could as well infer `upcastable: true` automatically if there's at least a `struct` inheriting it.
What do you think?

Reply:
Wait, `upcastable: true` is a property of the list
```yaml
  - name: Types
    doc: The type system
    # TODO: recognize this is upcastable
    list:
      type: SortedVector
      upcastable: true
      elementType: model::Type
```
which I use to tell wether the type of an element of the list is `UpcastablePointer<T>` or just `T`

Reply:
Ah sorry, my bad, ENOTENOUGHCONTEXT.
OK, then I'm happy with `upcastable` here, just drop the TODO.
Maybe we could make `upcastable: true` less C++-specific and say `accept_inheritors: true` or `accept_derived_structs: true`.